scripts/velocity.py

```diff
@@ -0,0 +1,452 @@
+        self.headers.update({'X-Authentication-Token': api_key})
+        self.private_api_key = private_api_key
+        self.hooks['response'] = [self._save_ratelimit_headers]
+        self.base_url = base_url
+        self.remaining = 0
+        self.reset_time = 0
+
+    def _save_ratelimit_headers(self, response: requests.Response, *_, **__):
+        """
+        Called when a response is generated from a :class:`requests.Request`.
+        Saves rate limiting header information used by :meth:`request`.
+
+        See https://github.com/ZenHubIO/API#api-rate-limit
+        """
+        limit = int(response.headers['X-RateLimit-Limit'])
+        used = int(response.headers['X-RateLimit-Used'])
+        self.remaining = limit - used
+        # ZenHub API documentation explicitly recommends using "their" time
+        # as indicated in the Date header
+        current_time = parsedate_to_datetime(response.headers['Date']).timestamp()
+        self.reset_time = current_time - int(response.headers['X-RateLimit-Reset'])
+
+    def request(self, method, url, *args, **kwargs) -> requests.Response:
+        url = urllib.parse.urljoin(self.base_url, url)
+        if self.remaining == 0:
+            # self.reset_time is sometimes negative, which makes no sense
+            # since ZH defines both operands
+            time.sleep(max(self.reset_time, 5))
+        return super(ZenHubAPI, self).request(method, url, *args, **kwargs)
+
+    def get_issue(self, repo_id: int, issue_num: int) -> requests.Response:
+        return self.get(f'/p1/repositories/{repo_id}/issues/{issue_num}')
+
+    def get_issue_history(self, repo_id: int, issue_num: int) -> requests.Response:
+        return self.get(f'/p1/repositories/{repo_id}/issues/{issue_num}/events')
+
+    def get_issue_history_private(self,
+                                  repo_id: int,
+                                  issue_num: int) -> requests.Response:
+        # ZenHub exposes issue-PR relationships only thru their private API
+        path = f'/v5/repositories/{repo_id}/issues/{issue_num}/events'
+        url = urllib.parse.urljoin(self.base_url, path)
+        r = requests.get(url, headers={'X-Authentication-Token': self.private_api_key})
+        r.raise_for_status()
+        return r
+
+
+class Issue(object):
+
+    def __init__(self, issue: github.Issue.Issue, zenhub: ZenHubAPI):
+        self.number = issue.number
+        self.title = issue.title
+        self.url = issue.html_url
+        self.labels = [label.name for label in issue.labels]
+        self.repository = issue.repository.full_name
+        self.repository_id = issue.repository.id
+        self.assignee = issue.assignee
+        self.assignees = issue.assignees
+        self.html_url = issue.html_url
+        self._zenhub: ZenHubAPI = zenhub
+
+    @functools.cached_property
+    def estimate(self) -> Optional[int]:
+        r = self._zenhub.get_issue(self.repository_id, self.number).json()
```

Comment:
You might want to turn `self._zenhub.get_issue(self.repository_id, self.number)` into property and cache the entire Zenhub issue as opposed to just the estimate.

Reply:
I decided to extra `self._zenhub.get_issue(self.repository_id, self.number)` into a property `zenhub_issue` for clarity, and moved the caching decorator to `ZenHubAPI.get_issue`.